### DISCUSSION TITLE
Converting .xopp files to .pdf and vice versa on windows (fast)

### DISCUSSION THREAD

**Discussion by leonausp** [xournalpp/xournalpp#3673] [Q&A] [answered]
Hello
I want to use xournalpp  for my study notes, but it is a real pain having to convert every single file per hand when finished. It would be really cool if i could at least convert all .xopp files in a folder to .pdf format and vice versa. It would save me so much time!
Does anyone know how to do that? (btw: I'm using windows)
It would really help me a lot.
Thanks!

**Comment by rolandlo** [comment, marked as answer]
First please note it is `xournalpp`, not `xournalapp`. 
Converting all `xopp`-files to `pdf` is easy in a console using bash:
```
for file in *.xopp; do echo Processing $file; xournalpp -p "${file%.*}"_export.pdf $file; done
```
Since you are on Windows, you don't have a console using bash right away, but you could install [MSYS2](https://www.msys2.org/) for that purpose. That will also solve your problem with not getting console output.

**Reply by leonausp** [reply]
Hello
Thanks very much for the reply, I changed it to *xounalpp* now.
So I installed MSYS2 now, but it's still not doing anything.
Do I maybe have to deinstall the .exe and download it this [this way](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md) for it to work?

**Reply by rolandlo** [reply]
No, you don't need to build from source. You should start xournalpp from the MSYS2 console. You can start it from the start menu.

**Reply by leonausp** [reply]
Ok, I did that.
Now I get:
````bash
for file in *.xopp; do echo Processing $file; xournalpp -p "${file%.*}"_export.pdf $file; done
Processing *.xopp
-bash: xournalpp: command not found
````

**Comment by rolandlo** [comment]
In the MSYS2 console you should change directory first. Assuming default installation folder this should work:
```
cd "/C/Program Files/Xournal++/bin"
./xournalpp.exe
```

**Reply by leonausp** [reply]
Ok, I did that.
Now I get:
````bash
for file in *.xopp; do echo Processing $file; xournalpp -p "${file%.*}"_export.pdf $file; done
Processing *.xopp
-bash: xournalpp: command not found
````

**Reply by rolandlo** [reply]
Ok, one thing I missed in that script: since you don't have the Xournal++ bin folder in your PATH, you need the absolute path for xournalpp:
> ```shell
> for file in *.xopp; do echo Processing $file; "/C/Program Files/Xournal++/bin/xournalpp.exe" -p "${file%.*}"_export.pdf $file; done
> ```
Secondly you need to execute the command in the folder where you have the .xopp files to convert. That is, cd into that folder instead of the Xournal++ bin folder.

**Reply by leonausp** [reply]
Thanks a lot! It worked, I was a bit dumb sorry, but now I'm happy it works :)
Do you have any idea how to make it recognize German characters (ä, ü, ö, etc.)? I tried it with a file named "Gemälde.xopp" and that threw an error, when I deleted the "ä" it worked flawlessly.

And sorry to bother you any further, but is there any chance that this could work backwards too, so that it can make a .xopp file for every pdf in a folder? (I tried replacing .xopp with pdf, and pdf with .opp but that didn't work)

**Reply by rolandlo** [reply]
On Linux special characters (like ä, ö, ü) don't seem to pose problems. On Windows it does for me as well. It might be something Xournal++ does wrong with the encoding. Feel free to open a bug report about that.
For a workaround you could just use a command to replace all special characters like in the chosen answer to [this question](https://serverfault.com/questions/348482/how-to-remove-invalid-characters-from-filenames)

**Comment by rolandlo** [comment]
Here is a little bash script `pdf2xopp.sh` I hacked together to create a `.xopp`-file from a `.pdf`-file: 
[pdf2xopp.zip](https://github.com/xournalpp/xournalpp/files/8123773/pdf2xopp.zip) 

Just extract it and make it executable. Remember to always check scripts before executing them blindly. You must have `mupdf-tools` installed to make it work. It is required for extracting the number of pages and width and height from the pages. On Ubuntu it can be installed via 
```shell
sudo apt install mupdf-tools
```

Usage: 
```shell
./pdf2xopp input.pdf output.xopp
```
For batch conversion:
```shell
for file in *.pdf; do echo Processing $file; ./pdf2xopp.sh "$file" "${file%.*}".xopp; done
```

**Reply by rolandlo** [reply]
In the nightly build there is now also a `--save` option:
```term
  -s, --save=XOPPFILE               Save xopp-file with the background PDF specified as FILE
```

So you don't need the script above (and don't need mupdf) and can write
```shell
"/C/Program Files/Xournal++/bin/xournalpp.exe" input.pdf --save=output.xopp
```
or for batch conversion:
```shell
for file in *.pdf; do echo Processing $file; "/C/Program Files/Xournal++/bin/xournalpp.exe" "$file" --save=${file%.*}".xopp; done
```

**Reply by flckf** [reply]
Thanks @rolandlo this is very helpful but is there any plan to include conversion from xopp to other formats? I need to batch convert from xopp files to png and this method doesn't seem to work. Thanks

**Reply by rolandlo** [reply]
@flckf What about the `-i` option? Suppose you are on Windows:
```term
for file in *.xopp; do echo Processing $file; "/C/Program Files/Xournal++/bin/xournalpp.exe" -i "${file%.*}".png $file; done
```

There are more options for .png, just run `xournalpp.exe --help-export`. 

>
```term
Advanced export options
  -p, --create-pdf=PDFFILE          Export FILE as PDF
  -i, --create-img=IMGFILE          Export FILE as image files (one per page)
                                 Guess the output format from the extension of IMGFILE
                                 Supported formats: .png, .svg
  --export-no-background            Export without background
                                 The exported file has transparent or white background,
                                 depending on what its format supports

  --export-no-ruling                Export without ruling
                                 The exported file has no paper ruling

  --export-layers-progressively     Export layers progressively
                                 In PDF export, Render layers progressively one by one.
                                 This results in N export pages per page with N layers,
                                 building up the layer stack progressively.
                                 The resulting PDF file can be used for a presentation.

  --export-range=RANGE              Only export the pages specified by RANGE (e.g. "2-3,5,7-")
                                 No effect without -p/--create-pdf or -i/--create-img
  --export-layer-range=RANGE        On export, only export the layers specified by RANGE (e.g. "2-3,5,7-")
                                 No effect without -p/--create-pdf or -i/--create-img
  --export-png-dpi=N                Set DPI for PNG exports. Default is 300
                                 No effect without -i/--create-img=foo.png
  --export-png-width=N              Set page width for PNG exports
                                 No effect without -i/--create-img=foo.png
                                 Ignored if --export-png-dpi is used
  --export-png-height=N             Set page height for PNG exports
                                 No effect without -i/--create-img=foo.png
                                 Ignored if --export-png-dpi or --export-png-width is used
```

**Reply by flckf** [reply]
Thanks for the quick reply-yes I am on windows.

sorry but a few stupid questions.. I have tried to open a command window in the xournalpp.exe folder and run the command "xournalpp.exe --help-export" which doesn't bring up the extra options.

I have also tried to run the command for a single file:

"C:/Program Files/Xournal++/bin/xournalpp.exe" C:\New folder\1.xopp --save-i=1.png 

Which doesn't work.. I assume I'm doing something obviously wrong?

Thanks again

**Reply by rolandlo** [reply]
On Windows, you don't get any output with the system provided cmd.exe. You would need to install [MSYS2](https://www.msys2.org/) and use that console to see the output.

In the command you used, you use the `--save` option which is only good to save a xopp-file for a PDF. As I wrote above, you need the `-i` option. For a single file:
```shell
"C:/Program Files/Xournal++/bin/xournalpp.exe" C:\New folder\1.xopp -i 1.png
```
It will write one png file per page.

**Reply by flckf** [reply]
ah right sorry I didn't realise the -i command needed MSYS2. I'll install it and try it out. Much appreciated

**Reply by rolandlo** [reply]
The `-i` option doesn't need MSYS2. Only getting command line output does.

**Reply by flckf** [reply]
ah right thanks- got it working! For some reason it doens't work on windows using code above- need to CD into the directory with the file and then run:  "C:/Program Files/Xournal++/bin/xournalpp.exe" 1.xopp -i 1.png

For some reason even having the path to the .xopp file in the command doesn't work. Just in case that helps anyone else..

Last question- for the batch convert, is that command something I can enter into a command window? I've tried the command as you have written it in a command window (cd to the relevant directory) as well as trying:

for file in C:\New folder\*.xopp; do echo Processing $file; "C:/Program Files/Xournal++/bin/xournalpp.exe" -i "${file%.*}".png $file; done

but I get a 'file was unexpected at this time' error... Any ideas?
